### PR TITLE
Dev reconnect

### PR DESCRIPTION
**Pull request by thecynic**
@cdheiser Here's an alternative to #23 Take a look?

### CODE REVIEW

File: pylutron/__init__.py [hunk 1]
```diff
@@ -93,12 +97,22 @@ def send(self, cmd):
...
     connection defaults like turning off the prompt, etc."""
-    self._telnet = telnetlib.Telnet(self._host)
-    self._telnet.read_until(LutronConnection.USER_PROMPT)
+    self._telnet = telnetlib.Telnet(self._host, timeout=2)  # 2 second timeout
+
+    # Ensure we know that connection goes away quickly, 5 sec
+    try:
+      to_attr = 18  # older pythons don't know about this one, so do our best
+      if hasattr(socket, 'TCP_USER_TIMEOUT'):
+        to_attr = getattr(socket, 'TCP_USER_TIMEOUT')
```
Comment (thecynic) [inline pylutron/__init__.py]: Unfortunately, this is linux only. Does anyone know if folks actually use HA on anything other than linux? If so I'll have to sort out a more platform independent solution, cobbled together with some keepalive options.

**Comment by cdheiser** [comment]
Yeah, they do.  There's even guides for installing on Windows and macOS.  I
don't know what the intersection of non-linux users and RadioRA2 users is
(probably zero, but maybe not?)

**Comment by thecynic** [comment]
Uploaded a replacement idle stuff that uses more standard options (should work on linux/mac), and looks like windows should work too: https://docs.microsoft.com/en-us/windows/win32/winsock/ipproto-tcp-socket-options

Can take feedback on the values chosen, as they were entirely arbitrary. This hsould close a dead connection in < 2 mins unless there outgoing commands, which will be faster.

**Comment by thecynic** [comment]
any opinions?

**Comment by JonGilmore** [comment]
how are you simulating a connection loss? I must be a rare case, but my connection rarely ever drops, or if it does, it reconnects with the original code already. I'm happy to toss it into my HA instance and run it for a week to make sure it doesn't totally break something else though

**Comment by thecynic** [comment]
I unplug the main repeater from power ;-)

**Comment by JonGilmore** [comment]
> I unplug the main repeater from power ;-)

derp, good call. I've tested this. From a functionality perspective, it works just the same as previous releases. But, it didn't solve the issue of a disconnect for me (at least inside of of Home Assistant).

**Comment by cdheiser** [comment]
Jon,

Do you end up with an exception/stacktrace in your home assistant logs when
this happens?

**Comment by JonGilmore** [comment]
> Jon, Do you end up with an exception/stacktrace in your home assistant logs when this happens?

Sorry - scratch my previous reply.... I was testing with the wrong code. It worked flawlessly for me just now while testing. I pulled power for 5 seconds on my main repeater and after a minute or so, here's a snippet from the logs:

`
2019-08-06 07:35:49 WARNING (Thread-2) [pylutron] Disconnected
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Connecting
2019-08-06 07:35:49 DEBUG (Thread-2) [pylutron] Sending: #MONITORING,12,2
2019-08-06 07:35:49 DEBUG (Thread-2) [pylutron] Sending: #MONITORING,255,2
2019-08-06 07:35:49 DEBUG (Thread-2) [pylutron] Sending: #MONITORING,3,1
2019-08-06 07:35:49 DEBUG (Thread-2) [pylutron] Sending: #MONITORING,4,1
2019-08-06 07:35:49 DEBUG (Thread-2) [pylutron] Sending: #MONITORING,5,1
2019-08-06 07:35:49 DEBUG (Thread-2) [pylutron] Sending: #MONITORING,6,1
2019-08-06 07:35:49 DEBUG (Thread-2) [pylutron] Sending: #MONITORING,8,1
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Connected
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Unknown cmd MONITORING (~MONITORING,12,2)
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Unknown cmd MONITORING (~MONITORING,255,2)
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Unknown cmd MONITORING (~MONITORING,3,1)
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Unknown cmd MONITORING (~MONITORING,4,1)
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Unknown cmd MONITORING (~MONITORING,5,1)
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Unknown cmd MONITORING (~MONITORING,6,1)
2019-08-06 07:35:49 INFO (Thread-2) [pylutron] Unknown cmd MONITORING (~MONITORING,8,1)
`

**Comment by JonGilmore** [comment]
fwiw, @thecynic @achatham, I've been running this for about a week in HASS with no issues. I tested the re-connect a couple times and it seems to have no issues. Anybody else successfully running it?

**Comment by JonGilmore** [comment]
Posting here because I think it's the best place... @grantalewis and I were chatting on Discord and I found what I think _may be_ a possible cause of his disconnects - does anybody know how Ra2 reacts when you re-use the same user/password for different integrations? From the [lutron docs](http://www.lutron.com/TechnicalDocumentLibrary/040249.pdf) on page 29, "Only one connection per login / password is allowed at a time". I know in his use case, he's using an OmniPro II with Ra2 as well as Home Assistant.

Thoughts?

**Comment by grantalewis** [comment]
I think it'd be worth investigating what in particular changed with pylutron after 0.84.6. I run that version for days on end without a hitch. But any version since exhibits the disconnect issue.

**Comment by JonGilmore** [comment]
> I think it'd be worth investigating what in particular changed with pylutron after 0.84.6. I run that version for days on end without a hitch. But any version since exhibits the disconnect issue.

From a cursory glance, it appears this is when HA switched from pylutron 0.1.0 to 0.2.0, I'll have to review changes as to what happened with those releases.

**Comment by thecynic** [comment]
Could you describe the disconnect issue? Stack trace? I don't think that
much changed on the connection side. Could your network be flaky in any way?

I'd like to get my most recent changes in and pulled into hass and see if
things improve.

Dima

**Comment by achatham** [comment]
I've seen cases where the lutron plugin is busted and I need to restart HA
to fix it, but I've never diagnosed it and haven't seen it in at least a
month or two.

**Comment by achatham** [comment]
And my PR to HA was accepted which bumps it to pylutron 0.22

**Comment by grantalewis** [comment]
> Could you describe the disconnect issue? > .

Description is here: https://github.com/home-assistant/home-assistant/issues/20348#issuecomment-515713889

As 0.84.6 is totally solid for me, I don't think it's my network.

**Comment by JonGilmore** [comment]
> I don't think that much changed on the connection side. Could your network be flaky in any way? I'd like to get my most recent changes in and pulled into hass and see if things improve.

I've reviewed things on the pylutron side and a bit on the HA side, and I agree that it doesn't look like much changed from a connection standpoint - which is what makes this so confusing....

**Comment by thecynic** [comment]
Ah I remember now. Dang, I did find a bug in the lutron light code in HASS
a while back that got introduced when someone did a refactor. I just never
sent it out :(

I'll send a PR tonight.

Dima

**Comment by JonGilmore** [comment]
> Ah I remember now. Dang, I did find a bug in the lutron light code in HASS a while back that got introduced when someone did a refactor. I just never sent it out :( I'll send a PR tonight. Dima

I'm curious as to what it is. I've never had any issues with disconnects, is it a bit of an edge case?

**Comment by thecynic** [comment]
The bug I fixed is the same one @JonGilmore fixed actually in https://github.com/home-assistant/home-assistant/pull/25592, though I think that solution will suffer from over-querying the MR. I can send a small fix for that.

**Comment by thecynic** [comment]
Ok, since there are no complaints about this patch and I've been running it for quite a while, I'm merging.

**Pull request merged**